Add CMSIS-DSP

stephanosio opened the pull request:
This PR adds the ARM CMSIS-DSP (digital signal processing) library integration to the Zephyr RTOS.

```
ext: hal: Add CMSIS-DSP library

This commit adds the initial port of the ARM CMSIS-DSP library for
use with the Zephyr RTOS.

The following is the list of the changes from the official CMSIS-DSP
library available in the upstream CMSIS distribution.

1. CMake build files (CMakeLists.txt) were re-written to be used as
  part of the Zephyr CMake build system.

2. The existing CMake-based configuration options were re-structured
  and re-written using Kconfig (menuconfig is fully supported).

There is no change from the upstream CMSIS-DSP distribution in terms
terms of the library implementation source code (*.c, *.h files).

Derived from the following commit in ARM-software/CMSIS_5:
b72a77cab226167186858a5fab0423fdc66e2193

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

This PR includes the commits from #21636.

Closes #21636 if merged before.

stephanosio committed:
ext: hal: Add CMSIS-DSP library

This commit adds the initial port of the ARM CMSIS-DSP library for
use with the Zephyr RTOS.

The following is the list of the changes from the official CMSIS-DSP
library available in the upstream CMSIS distribution.

1. CMake build files (CMakeLists.txt) were re-written to be used as
  part of the Zephyr CMake build system.

2. The existing CMake-based configuration options were re-structured
  and re-written using Kconfig (menuconfig is fully supported).

There is no change from the upstream CMSIS-DSP distribution in terms
terms of the library implementation source code (*.c, *.h files).

Derived from the following commit in ARM-software/CMSIS_5:
b72a77cab226167186858a5fab0423fdc66e2193

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'basic math' test application

This commit adds the test application for the CMSIS-DSP 'basic math'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: benchmarks: cmsis_dsp: Add 'basic math' benchmark application

This commit adds the benchmark application for the CMSIS-DSP 'basic
math' functions.

This benchmark application is loosely based on the C++ test suite
included in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'complex math' test application

This commit adds the test application for the CMSIS-DSP 'complex math'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'fast math' test application

This commit adds the test application for the CMSIS-DSP 'fast math'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'support' test application

This commit adds the test application for the CMSIS-DSP 'support'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'statistics' test application

This commit adds the test application for the CMSIS-DSP 'statistics'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'matrix' test application

This commit adds the test application for the CMSIS-DSP 'matrix'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'transform' test application

This commit adds the test application for the CMSIS-DSP 'transform'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'distance' test application

This commit adds the test application for the CMSIS-DSP 'distance'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'bayes' test application

This commit adds the test application for the CMSIS-DSP 'bayes'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'svm' test application

This commit adds the test application for the CMSIS-DSP 'svm'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'filtering' test application

This commit adds the test application for the CMSIS-DSP 'filtering'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ulfalizer commented on ext/hal/cmsis/DSP/Kconfig:
```diff
@@ -0,0 +1,509 @@
+
+config CMSIS_DSP_COMPLEXMATH
+	bool "Complex Math Functions"
+	imply CMSIS_DSP_FASTMATH
```
Just a note, but if `CMSIS_DSP_FASTMATH` starts out disabled in the initial configuration, then turning on `CMSIS_DSP_COMPLEXMATH` in menuconfig won't enable it.

Has to do with [this gotcha](https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig).

stephanosio replied:
> Just a note, but if `CMSIS_DSP_FASTMATH` starts out disabled in the initial configuration, then turning on `CMSIS_DSP_COMPLEXMATH` in menuconfig won't enable it.

That is indeed the case.

`CMSIS_DSP_FASTMATH` is "implied" instead of "selected" here because, depending on the "complex math" function being used, `CMSIS_DSP_FASTMATH` may not be required (and may be de-selected).

> Has to do with [this gotcha](https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig).

I reviewed the workarounds listed in the link above and none of them seem to provide an elegant solution.

With that being the case, is there a way to display a note/warning to the user if an implied symbol is not selected?

ulfalizer replied:
> With that being the case, is there a way to display a note/warning to the user if an implied symbol is not selected?

Besides checking the output, a warning could be added in `menuconfig`/`guiconfig` like this:

```
comment "Warning: You probably want foo enabled when bar is enabled"
	depends on BAR && !FOO
```

Was mostly of a note at least. Not sure how problematic it is in this case.

ulfalizer commented on ext/hal/cmsis/DSP/Kconfig:
```diff
@@ -0,0 +1,509 @@
+
+config CMSIS_DSP_NEON
+	bool "Neon Instruction Set"
+	# depends on CPU_CORTEX_A
```
Why commented out?

stephanosio replied:
`CPU_CORTEX_A` is commented out because it is currently not supported by Zephyr (there is an ongoing AArch64/ARMv8-A PR at the moment). It should be commented out when Cortex-A support is added in the future.

ulfalizer replied:
Maybe the comment could be tweaked to

```
# Cortex-A is not currently supported. Uncomment once it is.
# depends on CPU_CORTEX_A
```

ulfalizer commented on tests/lib/cmsis_dsp/filtering/Kconfig:
```diff
@@ -0,0 +1,13 @@
+config CMSIS_DSP_TEST_FILTERING_MISC
+	bool "Test: Filtering MISC"
+
+source "$ZEPHYR_BASE/Kconfig"
```
Can remove `$ZEPHYR_BASE` and do `source "Kconfig"`. The `srctree` environment variable will already be set to the Zephyr root.

Better to do `$(ZEPHYR_BASE)` btw, which uses the Kconfig preprocessor. `$ZEPHYR` is an older legacy syntax. Kconfiglib only supports it to be compatible with old Linux kernels.

stephanosio replied:
Done

ulfalizer commented on tests/lib/cmsis_dsp/matrix/Kconfig:
```diff
@@ -0,0 +1,25 @@
+config CMSIS_DSP_TEST_MATRIX_BINARY_F32
+	bool "Test: Matrix Binary F32"
+
+source "$ZEPHYR_BASE/Kconfig"
```
Ditto here (`source "Kconfig"`).

stephanosio replied:
Done

ulfalizer commented on tests/lib/cmsis_dsp/transform/Kconfig:
```diff
@@ -0,0 +1,22 @@
+config CMSIS_DSP_TEST_TRANSFORM_RF64
+	bool "Test: Transform RF64"
+
+source "$ZEPHYR_BASE/Kconfig"
```
Ditto here (`source "Kconfig"`).

stephanosio replied:
Done

ulfalizer commented on ext/hal/cmsis/Kconfig:
```diff
@@ -15,3 +15,14 @@ config HAS_CMSIS_CORE_R
+
+if HAS_CMSIS_DSP
+
+rsource "DSP/Kconfig"
```
`source` might make the path a bit more greppable, but not super important.

stephanosio replied:
Done

ulfalizer commented on ext/hal/cmsis/Kconfig:
```diff
@@ -15,3 +15,14 @@ config HAS_CMSIS_CORE_R
 
 endif
+
+menuconfig HAS_CMSIS_DSP
```
Wonder if this could be `select`ed from some SOC symbol. Does it make sense to turn it off in menuconfig?

stephanosio replied:
Unlike the CMSIS-Core component, which is an essential processor interface driver that is selected by the arch symbol, the CMSIS-DSP component is an optional library and therefore should be selected by the component/library/application that uses it.

Also it is worth noting that the CMSIS-DSP library can be used on the processor cores with and without the DSP extension.

ulfalizer replied:
Ah... might've been the naming that confused me. `HAS_...` to me sounds like it's describing something that's either available or not, and not an optional configurable feature.

ulfalizer reviewed (changes requested):
For `source "$ZEPHYR_BASE/Kconfig"` stuff. Rest is just some comments.

stephanosio committed:
boards: qemu_cortex_r5: Disable CMSIS-DSP tests

The CMSIS-DSP tests are disabled for the 'qemu_cortex_r5' platform
until the Cortex-R architecture port is more mature.

This commit should be reverted once the CMSIS-Core(R) integration
patches are merged (preliminary work available in #19698).

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
arch: arm: cortex_m: Add CPU DWT feature symbol

This commit adds a Kconfig symbol for specifying whether the SoC
implements the CPU DWT feature.

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: mps2: Disable CPU_CORTEX_M_HAS_DWT

This commit disables the CPU_CORTEX_M_HAS_DWT symbol for the MPS2
platforms (i.e. mps2_an385 and mps2_an521).

As far as the QEMU emulation goes, the DWT feature is not implemented
on these platforms and accessing the DWT registers results in a bus
fault.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
**Quick RFC**
(cc @nashif @ioannisg @MaureenHelm @galak)

Does anyone have any objections to making CMSIS a module?

ulfalizer commented on tests/lib/cmsis_dsp/filtering/src/biquad_f32.pat:
```diff
@@ -0,0 +1,267 @@
+static const uint32_t in_default_val[100] = {
+    0xbcf53bc7, 0x3e91ab21, 0x3f2b2c74, 0x3e9d1f6b, 
```
Nit: There's a trailing space on these lines. Probably from whatever generated them though.

ulfalizer reviewed (approved):
Kconfig stuff fixed.

ulfalizer commented:
Not sure what's going on with the internal CI error. Suspect it's something related to the [ci-pipelines](https://github.com/zephyrproject-rtos/ci-pipelines) repo.

@nashif

stephanosio commented:
NOTE: This PR will soon be reworked to be integrated into the new CMSIS west module.

ppryga commented:
@stephanosio are you going to mere this PR? Or will you drop and merge as part of west CMSIS module?

stephanosio commented:
> @stephanosio are you going to mere this PR? Or will you drop and merge as part of west CMSIS module?

@ppryga CMSIS-DSP itself will added as a new PR in zephyrproject-rtos/cmsis. This PR will only contain the tests and configs for the CMSIS-DSP.

I am waiting on ARM to release a new CMSIS version before proceeding with this, since this PR is based off the development branch and contains many unreleased features.

ppryga commented:
@stephanosio have you got any estimation when  the library may be available for use?

stephanosio commented:
> @stephanosio have you got any estimation when the library may be available for use?

I asked the CMSIS devs and they said the new CMSIS release (5.7.0) is expected to be ready some time in April, so probably in two months or so (I will get back to it as soon as they release 5.7.0).

ppryga commented:
@stephanosio 2 months is a bit long... can't we go with current release? How serious are the issues there?

stephanosio commented:
> @stephanosio 2 months is a bit long... can't we go with current release? How serious are the issues there?

If you want something that works right now, you can use this PR as is (it is fully functional on its own), though I cannot really say much about the functional integrity of the codebase since it was taken from the development branch mid-development.

ppryga commented:
@stephanosio 
I will take your advice into consideration.
Nevertheless, if that is functional then I do not understand the point to postpone merge until separate west module is ready? Any reasoning behind that? BTW, I do not think about project I'm working but also community that may find it useful.

stephanosio commented:
Superseded by #24307
